OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,293 @@
+    private:
+        ValueArrayDictionary _dictionary;
+    }; // ValueDictionary
+
+public:
+    using TableMetaData       = ValueDictionary;
+    using DependentsMetaData  = ValueArrayDictionary;
+    using IndependentMetaData = ValueDictionary;
+
+    AbstractDataTable()                                      = default;
+    AbstractDataTable(const AbstractDataTable&)              = default;
+    AbstractDataTable(AbstractDataTable&&)                   = default;
+    AbstractDataTable& operator=(const AbstractDataTable&)   = default;
+    AbstractDataTable& operator=(AbstractDataTable&&)        = default;
+    virtual std::unique_ptr<AbstractDataTable> clone() const = 0;
+    virtual ~AbstractDataTable()                             = default;
+
+    /** Get metadata associated with the table.                               */
+    const TableMetaData& getTableMetaData() const {
+        return _tableMetaData;
+    }
+
+    /** Update metadata associated with the table.                            */
+    TableMetaData& updTableMetaData() {
+        return _tableMetaData;
+    }
+
+    /** Get number of rows.                                                   */
+    size_t getNumRows() const {
+        return computeNumRows();
+    }
+
+    /** Get number of columns.                                                */
+    size_t getNumColumns() const {
+        return computeNumColumns();
+    }
+    
+    /** Get metadata associated with the independent column.                  */
+    const IndependentMetaData& getIndependentMetaData() const {
+        return _independentMetaData;
+    }
+    
+    /** Set metadata associated with the independent column.                  */
+    void 
+    setIndependentMetaData(const IndependentMetaData& independentMetaData) {
+        _independentMetaData = independentMetaData;
+        validateIndependentMetaData();
+    }
+
+    /** Get metadata associated with the dependent columns.                   */
+    const DependentsMetaData& getDependentsMetaData() const {
+        return _dependentsMetaData;
+    }
+    
+    /** Set metadata associated with the dependent columns.                   */
+    void 
+    setDependentsMetaData(const DependentsMetaData& dependentsMetaData) {
+        _dependentsMetaData = dependentsMetaData;
+        validateDependentsMetaData();
+    }
+
+protected:
+    virtual size_t computeNumRows() const            = 0;
+    virtual size_t computeNumColumns() const         = 0;
+    virtual void validateIndependentMetaData() const = 0;
+    virtual void validateDependentsMetaData() const  = 0;
+
+    TableMetaData       _tableMetaData;
+    DependentsMetaData  _dependentsMetaData;
+    IndependentMetaData _independentMetaData;
+}; // AbstractDataTable
+
+
+/** DataTable_ is a in-memory storage container for data with support for 
+holding metadata. Data contains an independent column and a set of dependent 
+columns. The type of the independent column can be configured using ETX (
```

Comment:
It is a bit late to be discussing meta data here since `DataTable_` derives from `AbstractDataTable` where the description of meta data should be complete. 

I think it would be helpful to say that `DataTable_` is used to construct concrete DataTables of varying independent (`ETX`) and dependent (`ETY`) numerical data types.